lgc/patch/MeshTaskShader.cpp

```diff
@@ -0,0 +1,453 @@
+  }
+
+  // Construct ".emitTaskMeshs" block
+  {
+    m_builder->SetInsertPoint(emitMeshsBlock->getTerminator());
+
+    //
+    // Collect task statistics info
+    //
+    auto &computeMode =
+        m_pipelineState->getShaderModes()->getComputeShaderMode(); // Task shader is actually a compute shader
+    const uint64_t numTaskThreads =
+        computeMode.workgroupSizeX * computeMode.workgroupSizeY * computeMode.workgroupSizeZ;
+
+    auto meshPipeStatsBufPtr = m_pipelineSysValues.get(entryPoint)->getMeshPipeStatsBufPtr();
+    auto meshPipeStatsBufEntryPtr = m_builder->CreateGEP(
+        m_builder->getInt8Ty(), meshPipeStatsBufPtr, m_builder->getInt32(offsetof(MeshPipeStatsEntry, numTaskThreads)));
+    meshPipeStatsBufEntryPtr = m_builder->CreateBitCast(meshPipeStatsBufEntryPtr,
+                                                        PointerType::get(m_builder->getInt64Ty(), ADDR_SPACE_GLOBAL));
+
+    // NOTE: LLVM backend will try to apply atomics optimization. But here, we only have one active thread to execute
+    // the global_atomic_add instruction. Thus, the optimization is completely unnecessary. To avoid this, we try to
+    // move the added value to VGPR to mark it as "divergent".
+    Value *valueToAdd = UndefValue::get(FixedVectorType::get(m_builder->getInt32Ty(), 2));
+    valueToAdd = m_builder->CreateInsertElement(valueToAdd, convertToDivergent(m_builder->getInt32(numTaskThreads)),
+                                                static_cast<uint64_t>(0));
+    valueToAdd =
+        m_builder->CreateInsertElement(valueToAdd, convertToDivergent(m_builder->getInt32(numTaskThreads >> 32)), 1);
+    valueToAdd = m_builder->CreateBitCast(valueToAdd, m_builder->getInt64Ty());
+
+    m_builder->CreateAtomicRMW(AtomicRMWInst::Add, meshPipeStatsBufEntryPtr, valueToAdd, MaybeAlign(),
+                               AtomicOrdering::Monotonic, SyncScope::System);
+
+    //
+    // Write draw data
+    //
+
+    // Set X dimension to 0 if any of X, Y, Z dimension is 0:
+    //   groupCountX = min(groupCountY, groupCountZ) == 0 ? 0 : groupCountX
+    auto minGroupCountYZ =
+        m_builder->CreateIntrinsic(Intrinsic::umin, groupCountY->getType(), {groupCountY, groupCountZ});
+    groupCountX = m_builder->CreateSelect(m_builder->CreateICmpEQ(minGroupCountYZ, m_builder->getInt32(0)),
+                                          m_builder->getInt32(0), groupCountX);
+
+    auto drawDataRingBufDesc = m_pipelineSysValues.get(entryPoint)->getTaskDrawDataRingBufDesc();
+    auto drawDataRingEntryOffset = getDrawDataRingEntryOffset(entryPoint);
+
+    // Draw data (<4 x i32>) = <groupCountX, groupCountY, groupCountZ, readyBit>
+    Value *drawData = UndefValue::get(FixedVectorType::get(m_builder->getInt32Ty(), 4));
+    drawData = m_builder->CreateInsertElement(drawData, groupCountX, static_cast<uint64_t>(0));
+    drawData = m_builder->CreateInsertElement(drawData, groupCountY, 1);
```

Comment:
nit: why do we use static_cast to in only one of the calls?

Reply:
For reasons, C++ decides to cast a literal 0 into a `nullptr`, making the `CreateInsertElement` call ambiguous between taking a literal int and a pointer. This only happens for 0, but not for 1. Instead of a cast, using `0ul` should also work.